hydra-plutus/src/Hydra/Contract/Head.hs

```diff
@@ -248,46 +257,41 @@ checkClose ::
   ScriptContext ->
   [Party] ->
   BuiltinByteString ->
-  SnapshotNumber ->
-  BuiltinByteString ->
   [Signature] ->
   ContestationPeriod ->
   CurrencySymbol ->
   Bool
-checkClose ctx parties initialUtxoHash snapshotNumber closedUtxoHash sig cperiod headPolicyId =
+checkClose ctx parties initialUtxoHash sig cperiod headPolicyId =
   hasBoundedValidity
+    && checkDeadline
     && checkSnapshot
     && mustBeSignedByParticipant ctx headPolicyId
     && hasST headPolicyId outValue
  where
-  hasBoundedValidity = traceIfFalse "hasBoundedValidity check failed" $ tMax - tMin <= cp
+  hasBoundedValidity =
+    traceIfFalse "hasBoundedValidity check failed" $
+      tMax - tMin <= cp
 
   outValue =
     maybe mempty (txOutValue . txInInfoResolved) $ findOwnInput ctx
 
+  (closedSnapshotNumber, closedUtxoHash, closedContestationDeadline) =
+    -- XXX: fromBuiltinData is super big (and also expensive?)
+    case fromBuiltinData @DatumType $ getDatum (continuingDatum ctx) of
```

Comment:
An interesting CIP which is related: https://github.com/cardano-foundation/CIPs/pull/455